src/v/arch/manifest.cc

```diff
@@ -0,0 +1,206 @@
+
+bool manifest::contains(const segment_name& obj) const {
+    return _segments.count(obj) != 0;
+}
+
+bool manifest::add(const segment_name& key, const segment_meta& meta) {
+    auto [it, ok] = _segments.insert(std::make_pair(key, meta));
+    return ok;
+}
+
+std::optional<std::reference_wrapper<const manifest::segment_meta>>
+manifest::get(const segment_name& key) const {
+    auto it = _segments.find(key);
+    if (it == _segments.end()) {
+        return std::nullopt;
+    }
+    return std::cref(it->second);
+}
+
+std::insert_iterator<manifest::segments_map> manifest::get_insert_iterator() {
+    return std::inserter(_segments, _segments.begin());
+}
+
+manifest manifest::difference(const manifest& remote_set) const {
+    if (_ntp != remote_set._ntp && _rev != remote_set._rev) {
+        throw manifest_error("NTPs doesn't match");
+    }
+    manifest result(_ntp, _rev);
+    std::set_difference(
+      begin(),
+      end(),
+      remote_set.begin(),
+      remote_set.end(),
+      result.get_insert_iterator());
+    return result;
+}
+
+ss::future<> manifest::update(ss::input_stream<char>&& is) {
+    iobuf result;
+    auto os = make_iobuf_ref_output_stream(result);
+    co_await ss::copy(is, os);
+    iobuf_istreambuf ibuf(result);
+    std::istream stream(&ibuf);
+    ptree m;
+    boost::property_tree::json_parser::read_json(stream, m);
+    update(m);
+    co_return;
+}
+
+void manifest::update(const manifest::ptree& m) {
+    // ntp from json
+    auto ns = model::ns(m.get<ss::sstring>("namespace"));
+    auto tp = model::topic(m.get<ss::sstring>("topic"));
+    auto pt = model::partition_id(m.get<int32_t>("partition"));
+    _rev = model::revision_id(m.get<int64_t>("revision"));
+    _ntp = model::ntp(ns, tp, pt);
+    segments_map tmp;
+    if (m.count("segments") != 0) {
+        for (const ptree::value_type& it : m.get_child("segments")) {
+            auto coffs = it.second.get<int64_t>("committed_offset");
+            auto boffs = it.second.get<int64_t>("base_offset");
+            segment_meta meta{
+              .is_compacted = it.second.get<bool>("is_compacted"),
+              .size_bytes = it.second.get<size_t>("size_bytes"),
+              .base_offset = model::offset(boffs),
+              .committed_offset = model::offset(coffs),
+              .is_deleted_locally = it.second.get<bool>("deleted"),
+            };
+            tmp.insert(std::make_pair(segment_name(it.first.data()), meta));
+        }
+    }
+    std::swap(tmp, _segments);
+}
+
+std::tuple<ss::input_stream<char>, size_t> manifest::serialize() const {
+    iobuf serialized;
+    iobuf_ostreambuf obuf(serialized);
+    std::ostream os(&obuf);
+    serialize(os, false);
+    size_t size_bytes = serialized.size_bytes();
+    return std::make_tuple(
+      make_iobuf_input_stream(std::move(serialized)), size_bytes);
+}
+
+void manifest::serialize(std::ostream& out, bool pretty) const {
+    ptree m;
```

Comment:
this is crazy! how come ptree has no namespace? shouldn't it be `boost::container::ptree` or smth like that?

Reply:
ah you defined it in the header.